### PR TITLE
feat: sync manifest from TDP

### PR DESCRIPTION
**Pull request by anchenyi**
https://msazure.visualstudio.com/Microsoft%20Teams%20Extensibility/_workitems/edit/27586844

### CODE REVIEW

File: packages/fx-core/resource/package.nls.json [hunk 1]
```diff
@@ -577,6 +581,18 @@
   "core.uninstallQuestion.tdpOption": "App registration",
   "core.uninstallQuestion.botOption": "Bot framework registration",
   "core.uninstallQuestion.projectPath": "Project path",
+  "core.syncManifest.projectPath": "Project path",
+  "core.syncManifest.env": "Target Teams Toolkit Environment",
+  "core.syncManifest.teamsAppId": "Teams App ID (optional)",
+  "core.syncManifest.addWarning": "There are add changes in the manifest template, please manually update the local manifest. Diff Path: %s. New Value %s.",
```
Comment (HuihuiWu-Microsoft) [inline packages/fx-core/resource/package.nls.json:587]: add changes -> added changes or add quotes?

Comment (anchenyi) [reply packages/fx-core/resource/package.nls.json]: Changed to a dfiferent expression.Thanks for pointing out.

**Review by HuihuiWu-Microsoft — DISMISSED** [review]

File: packages/fx-core/resource/package.nls.json [hunk 2]
```diff
@@ -134,6 +134,7 @@
   "plugins.appstudio.validationFailedNotice": "Manifest Validation is unsuccessful!",
   "plugins.appstudio.validateManifest.progressBar.message": "Validating manifest...",
   "plugins.appstudio.validateAppPackage.progressBar.message": "Validating app package...",
+  "plugins.appstudio.syncManifestFailedNotice": "Failed to sync manifest!",
```
Comment (supkasar) [inline packages/fx-core/resource/package.nls.json:137]: Unable to sync manifest!

File: packages/fx-core/resource/package.nls.json [hunk 3]
```diff
@@ -577,6 +581,18 @@
   "core.uninstallQuestion.tdpOption": "App registration",
   "core.uninstallQuestion.botOption": "Bot framework registration",
   "core.uninstallQuestion.projectPath": "Project path",
+  "core.syncManifest.projectPath": "Project path",
+  "core.syncManifest.env": "Target Teams Toolkit Environment",
+  "core.syncManifest.teamsAppId": "Teams App ID (optional)",
+  "core.syncManifest.addWarning": "There are new properties added to the manifest template, please manually update the local manifest. Diff Path: %s. New Value %s.",
```
Comment (supkasar) [inline packages/fx-core/resource/package.nls.json:587]: New properties added to the manifest template. Manually update the local manifest. Diff Path: %s. New Value %s.

File: packages/fx-core/resource/package.nls.json [hunk 4]
```diff
@@ -577,6 +581,18 @@
...
+  "core.syncManifest.deleteWarning": "There are delete changes in the manifest template, please manually update the local manifest. Diff Path: %s. Old Value: %s.",
```
Comment (supkasar) [inline packages/fx-core/resource/package.nls.json:588]: Something was deleted from the manifest template. Manually update the local manifest. Diff Path: %s. Old Value: %s.

File: packages/fx-core/resource/package.nls.json [hunk 5]
```diff
@@ -577,6 +581,18 @@
...
+  "core.syncManifest.editKeyConflict": "There is a conflict in the placeholder variable in the new manifest. Please manually update the local manifest. Variable name: %s, value 1: %s, value 2: %s.",
```
Comment (supkasar) [inline packages/fx-core/resource/package.nls.json:589]: Conflict in placeholder variable in the new manifest. Manually update the local manifest. Variable name: %s, value 1: %s, value 2: %s.

File: packages/fx-core/resource/package.nls.json [hunk 6]
```diff
@@ -577,6 +581,18 @@
...
+  "core.syncManifest.editNonVarPlaceholder": "The new manifest contains changes that are not placeholder variables. Please manually update the local manifest. Old value: %s. New Value: %s.",
```
Comment (supkasar) [inline packages/fx-core/resource/package.nls.json:590]: The new manifest has non-placeholder changes. Manually update your local manifest. Old value: %s. New value: %s.

File: packages/fx-core/resource/package.nls.json [hunk 7]
```diff
@@ -577,6 +581,18 @@
...
+  "core.syncManifest.editNotMatch": "Cannot match the value with placeholders in the template. Please manually update the local manifest. Template value: %s. New Value: %s.",
```
Comment (supkasar) [inline packages/fx-core/resource/package.nls.json:591]: Value doesn't match the template placeholders. Manually update the local manifest. Template value: %s. New Value: %s.

File: packages/fx-core/resource/package.nls.json [hunk 8]
```diff
@@ -577,6 +581,18 @@
...
+  "core.syncManifest.updateEnvSuccess": "Sucessfully updated the %s environment file. New values: %s",
```
Comment (supkasar) [inline packages/fx-core/resource/package.nls.json:592]: %s environment file updated successfully. New values: %s

File: packages/fx-core/resource/package.nls.json [hunk 9]
```diff
@@ -577,6 +581,18 @@
...
+  "core.syncManifest.success": "Sucessfully synced manifest to environment: %s.",
```
Comment (supkasar) [inline packages/fx-core/resource/package.nls.json:593]: Manifest successfully synced to environment: %s.

File: packages/fx-core/resource/package.nls.json [hunk 10]
```diff
@@ -577,6 +581,18 @@
...
+  "core.syncManifest.noDiff": "Your manifest file is already up-to-date. Sync completed.",
+  "core.syncManifest.saveManifestSuccess": "Successfully saved the manifest file to %s.",
```
Comment (supkasar) [inline packages/fx-core/resource/package.nls.json:595]: Manifest file saved to %s successfully.

File: packages/fx-core/resource/package.nls.json (hunk 3 above)
Comment (anchenyi) [reply packages/fx-core/resource/package.nls.json:587]: For this warning, users need to update the manifest by their own. What about adding a "Please" before the "Manually update the local manifest"?

File: packages/fx-core/resource/package.nls.json [hunk 11]
```diff
@@ -195,6 +196,9 @@
   "error.appstudio.noManifestId": "Invalid ID found in manifest find.",
   "error.appstudio.validateFetchSchemaFailed": "Unable to get schema from %s, message: %s",
   "error.appstudio.validateSchemaNotDefined": "Manifest schema is not defined",
+  "error.appstudio.syncManifestInvalidInput": "Input is invalid. Project path and env should not be empty.",
+  "error.appstudio.syncManifestNoTeamsAppId": "Cannot load a Teams app ID from the env file.",
```
Comment (supkasar) [inline packages/fx-core/resource/package.nls.json:200]: Unable to load Teams app ID from the env file.

**Review by supkasar — APPROVED** [review]

**Review by jayzhang — APPROVED** [review]

File: packages/fx-core/src/question/other.ts [hunk 12]
```diff
@@ -1066,3 +1066,48 @@ function oauthClientSecretQuestion(): TextInputQuestion {
...
+    },
+    children: [
+      {
+        data: {
+          type: "folder",
+          name: QuestionNames.ProjectPath,
+          title: getLocalizedString("core.syncManifest.projectPath"),
+          cliDescription: "Project Path",
+          placeholder: "./",
+          default: "./",
```
Comment (yuqizhou77) [inline packages/fx-core/src/question/other.ts:1083]: should the default be the same as default folder when TTK creates a new project?

File: packages/fx-core/src/question/other.ts [hunk 13]
```diff
@@ -1066,3 +1066,48 @@ function oauthClientSecretQuestion(): TextInputQuestion {
...
+        },
+        condition: () => {
+          return true;
```
Comment (yuqizhou77) [inline packages/fx-core/src/question/other.ts:1086]: Is the conditional required here? Looks like the question should always be shown, so may not need to set the condition?

File: packages/fx-core/src/question/other.ts [hunk 14]
```diff
@@ -1066,3 +1066,48 @@ function oauthClientSecretQuestion(): TextInputQuestion {
...
+          title: getLocalizedString("core.syncManifest.env"),
+          cliDescription: "Target Teams Toolkit Environment",
+        },
+        condition: () => {
+          return true;
+        },
+      },
+      {
+        data: {
+          type: "text",
```
[the remark this answers is not in the document]
Comment (anchenyi) [reply packages/fx-core/src/question/other.ts:1102]: Besides the uri handler, users can also use the command (ctrl +shift + P), so we need the teamsApp ID question for this case.

File: packages/fx-core/src/question/other.ts (hunk 13 above)
Comment (anchenyi) [reply packages/fx-core/src/question/other.ts:1086]: Yes, you are right. I will remove the condition here.

**Review by yuqizhou77 — APPROVED** [review]

File: packages/fx-core/src/question/other.ts (hunk 12 above)
Comment (anchenyi) [reply packages/fx-core/src/question/other.ts:1083]: Thanks for pointing out. Will change this.

**Review by HuihuiWu-Microsoft — APPROVED** [review]

File: packages/vscode-extension/src/handlers/manifestHandlers.ts [hunk 15]
```diff
@@ -33,6 +35,17 @@ export async function validateManifestHandler(args?: any[]): Promise<Result<null
   return await runCommand(Stage.validateApplication, inputs);
 }
 
+export async function syncManifestHandler(...args: any[]): Promise<Result<null, FxError>> {
+  ExtTelemetry.sendTelemetryEvent(TelemetryEvent.SyncManifest, getTriggerFromProperty(args));
```
[the remark this answers is not in the document]
Comment (anchenyi) [reply packages/vscode-extension/src/handlers/manifestHandlers.ts:39]: Thanks for pointinh out. I will update this in another PR.

**Review by wenytang-ms — APPROVED** [review]

**Pull request merged**